src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,214 @@
+        /// "take the compilation this other thing produced, and pretend the generator gave this content, even if it wouldn't."
+        /// </summary>
+        private class GeneratedFileReplacingCompilationTracker : ICompilationTracker
+        {
+            private readonly ICompilationTracker _underlyingTracker;
+            private readonly SourceGeneratedDocumentState _replacedGeneratedDocumentState;
+
+            /// <summary>
+            /// The lazily-produced compilation that has the generated document updated. This is initialized by call to
+            /// <see cref="GetCompilationAsync"/>.
+            /// </summary>
+            [DisallowNull]
+            private Compilation? _compilationWithReplacement;
+
+            public GeneratedFileReplacingCompilationTracker(ICompilationTracker underlyingTracker, SourceGeneratedDocumentState replacementDocumentState)
+            {
+                _underlyingTracker = underlyingTracker;
+                _replacedGeneratedDocumentState = replacementDocumentState;
+            }
+
+            public bool HasCompilation => _underlyingTracker.HasCompilation;
+
+            public ProjectState ProjectState => _underlyingTracker.ProjectState;
+
+            public ICompilationTracker Clone()
+            {
+                return new GeneratedFileReplacingCompilationTracker(_underlyingTracker.Clone(), _replacedGeneratedDocumentState);
+            }
+
+            public bool ContainsAssemblyOrModuleOrDynamic(ISymbol symbol, bool primary)
+            {
+                if (_compilationWithReplacement == null)
+                {
+                    // We don't have a compilation yet, so this couldn't have came from us
+                    return false;
+                }
+                else
+                {
+                    return UnrootedSymbolSet.Create(_compilationWithReplacement).ContainsAssemblyOrModuleOrDynamic(symbol, primary);
+                }
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(Func<string, bool> predicate, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                // This only needs to be implemented if a feature that operates from a source generated file needs to search for declarations
+                // with other names; those APIs are only used by certain code fixes which isn't a need for now.
+                throw new NotImplementedException();
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(string name, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                // This only needs to be implemented if a feature that operates from a source generated file needs to search for declarations
+                // with other names; those APIs are only used by certain code fixes which isn't a need for now.
+                throw new NotImplementedException();
+            }
```

Comment:
so, i would prefer we impelment this similar to the main tracker.  if we don't have the compilation, we return null, otherwise we query that compilation.